Mathlib/Algebra/Operad/Operad.lean

```diff
@@ -0,0 +1,136 @@
+/-
+Copyright (c) 2024 Alex Meiburg. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Meiburg
+-/
+import Mathlib.Algebra.Operad.Basic
+import Mathlib.Algebra.Operad.Perm
+import Mathlib.GroupTheory.Perm.Basic
+
+/-! Defines `Operad` and `SymmOperad`. An `Operad` is a ℕ-graded set of operations that are
+  closed under composition, with an identity, that obey the "natural" commutative and associative
+  constraints - the ones satisfied by composition of arbitrary multi-argument functions.
+
+  A `SymmOperad` is an `Operad` equipped with a `MulAction` by `Perm (Fin n)` for all n, that
+  rearrange the arguments in the natural way.
+-/
+
+section Operad
+
+/-- An `Operad` is a ℕ-graded set of operations that are closed under composition and
+  with an identity. The composition operation has to be appropriately associative and
+  commutative:
+  - Composing a into b and then into c is assocative as to the order of composition.
+  - Composing a and b into two different arguments of c is commutative as to which is
+    composed in first.
+
+  Operads are sometimes defined instead with "superposition" as the basic operation, where
+  all n arguments are composed at once instead of just one. This is an equivalent structure,
+  as superposition can be built from progressively composing each in and using identity in
+  the other arguments. This notion of superposition is rather unwieldly in Lean, since
+  the *type* of the result is the sum of the arities of all the arguments, and associativity
```

Comment:
I'm not sure that it would be harder: you're currently doing a lot of suffering with `PermFinPadAt` and friends.

I'm also a bit worried that this approach is harder to generalize to the coloured case (where at least in my experience all the interesting examples lie?). But I guess we have to get started.

I think the material in Operad/Perm could be made a lot nicer.